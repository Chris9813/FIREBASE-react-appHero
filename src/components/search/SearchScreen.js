import React, { useMemo } from 'react'
import queryString from "query-string";
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {

    const location = useLocation();
    const {q = ""} = queryString.parse(location.search)

    const [values, handleInnputChange] = useForm({
        name: q,
    })
    const {name} = values
    const memo = useMemo(() => getHeroesByName(q), [q])
    const handleSearch = (e) => {
        e.preventDefault()
        history.push(`?q=${name}`);
    }
    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className ="col-5">
                    <form onSubmit = {handleSearch}>
                        <input  
                        type = "text"
                        placeholder ="Find your hero"
                        className ="form-control"
                        onChange = {handleInnputChange}
                        name = "name"
                        value = {name}
                        autoComplete = "off"
                        />
                    <button 
                    onClick = {handleSearch}
                    className = "btn m-1 btn-block btn-outline-primary">
                        Search...
                    </button>
                    </form>
                </div>
                <div className ="col-7">
                    <h4>Results</h4>
                    <hr/>
                    {q === "" && 
                                <div id = "infoIni" className = "alert alert-info">
                                search a hero
                                </div>}
                    {(q !== "" && memo.length === 0)
                            &&
                            <div className = "alert alert-warning">
                            There is no a hero {q}, is a bich, search anoter thing
                            </div>}
                    {
                        memo.map(hero => {
                            return <HeroCard 
                            key = {hero.id}
                            {...hero}
                            />
                        })
                    }
                </div>
            </div>
        </div>
    )
}
