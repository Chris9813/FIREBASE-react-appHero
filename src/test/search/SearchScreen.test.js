import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { SearchScreen } from "../../components/search/SearchScreen"


describe('pruebas en el componente SearchScreen ', () => {
    
    test('debe mostrarse correctamente con valores por defecto ', () => {
        const wrapper = mount(
        <MemoryRouter initialEntries = {["/search"]}>
            <Route path = "/search" component = {SearchScreen}/>
        </MemoryRouter>        
        )
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("search a hero")
    })
    
    test('debe mostrar  batman y el input con el valor del queryString ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries = {["/search?q=batman"]}>
                <Route path = "/search" component = {SearchScreen}/>
            </MemoryRouter>        
            )
        expect( wrapper.find("input").prop("value")).toBe("batman") 
        expect( wrapper ).toMatchSnapshot();
        console.log(wrapper.find("input").prop("value"));
    })

    test('debe de mostrar un error si no se encuentra el hero ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries = {["/search?q=batman12"]}>
                <Route path = "/search" component = {SearchScreen}/>
            </MemoryRouter>        
            )
        expect( wrapper.find("input").prop("value")).toBe("batman12") 
        expect( wrapper ).toMatchSnapshot();
        console.log(wrapper.find("input").prop("value"));
        expect(wrapper.find(".alert-warning").text().trim()).toBe("There is no a hero batman12, is a bich, search anoter thing");
    })


    test('debe de llamar el push del history ', () => {
        const history = {
            push: jest.fn(),
        };
        const wrapper = mount(
            <MemoryRouter initialEntries = {["/search?q=batman"]}>
                <Route path = "/search" 
                component = {() => <SearchScreen history = {history}/>}/>
            </MemoryRouter>        
            )
        
        wrapper.find("input").simulate("change", {
            target : {
                name: "name",
                value: "batman",
            }
        });

        wrapper.find("form").prop("onSubmit")({
            preventDefault(){}
        })

        expect( history.push ).toHaveBeenCalledWith(`?q=batman`)
    })

    

})
