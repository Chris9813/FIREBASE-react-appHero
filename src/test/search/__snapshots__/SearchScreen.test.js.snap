// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el componente SearchScreen  debe de mostrar un error si no se encuentra el hero  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman12"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          onClick={[Function]}
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-warning"
      >
        There is no a hero 
        batman12
        , is a bich, search anoter thing
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en el componente SearchScreen  debe mostrar  batman y el input con el valor del queryString  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          onClick={[Function]}
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="row row-cols-2"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="card-body"
        >
          <img
            alt="{superhero}"
            className="card-img-top"
            src="./assets/heroes/dc-batman.jpg"
          />
        </div>
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            Batman
          </h5>
          <p
            className="card-text-center"
          >
            Bruce Wayne
          </p>
          <p
            className="card-text"
          >
            <small>
              Detective Comics #27
            </small>
          </p>
          <a
            className="link-info"
            href="/hero/dc-batman"
            onClick={[Function]}
          >
            Mas...
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en el componente SearchScreen  debe mostrarse correctamente con valores por defecto  1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          onClick={[Function]}
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
        id="infoIni"
      >
        search a hero
      </div>
    </div>
  </div>
</div>
`;
